internal/public/endpoints/vehicle.go

```diff
@@ -0,0 +1,207 @@
+package endpoints
+
+import (
+	"context"
+	"math"
+
+	"github.com/jackc/pgx/v5/pgtype"
+	"github.com/jamespfennell/transiter/internal/convert"
+	"github.com/jamespfennell/transiter/internal/gen/api"
+	"github.com/jamespfennell/transiter/internal/gen/db"
+	"github.com/jamespfennell/transiter/internal/public/errors"
+)
+
+func ListVehicles(ctx context.Context, r *Context, req *api.ListVehiclesRequest) (*api.ListVehiclesReply, error) {
+	system, err := getSystem(ctx, r.Querier, req.SystemId)
+	if err != nil {
+		return nil, err
+	}
+
+	numVehicles := r.EndpointOptions.MaxVehiclesPerRequest
+	if numVehicles <= 0 {
+		numVehicles = math.MaxInt32 - 1
+	}
+	if req.Limit != nil && *req.Limit < numVehicles {
+		numVehicles = *req.Limit
+	}
+	var firstID string
+	if req.FirstId != nil {
+		firstID = *req.FirstId
+	}
+
+	var dbVehicles []db.ListVehiclesRow
+	if req.GetSearchMode() == api.ListVehiclesRequest_DISTANCE {
+		if req.Latitude == nil || req.Longitude == nil || req.MaxDistance == nil {
+			return nil, errors.NewInvalidArgumentError("latitude, longitude, and max_distance are required when using DISTANCE search_mode")
+		}
+		if firstID != "" {
+			return nil, errors.NewInvalidArgumentError("first_id can not be used when using DISTANCE search_mode")
+		}
+		dbVehiclesGeo, err := r.Querier.ListVehicles_Geographic(ctx, db.ListVehicles_GeographicParams{
+			SystemPk:    system.Pk,
+			NumVehicles: numVehicles,
+			Latitude:    convert.Gps(req.Latitude),
+			Longitude:   convert.Gps(req.Longitude),
+			MaxDistance: convert.Gps(req.MaxDistance),
+		})
+
+		for _, dbVehicle := range dbVehiclesGeo {
+			dbVehicles = append(dbVehicles, db.ListVehiclesRow{
+				ID:                  dbVehicle.ID,
+				Label:               dbVehicle.Label,
+				LicensePlate:        dbVehicle.LicensePlate,
+				CurrentStatus:       dbVehicle.CurrentStatus,
+				Latitude:            dbVehicle.Latitude,
+				Longitude:           dbVehicle.Longitude,
+				Bearing:             dbVehicle.Bearing,
+				Odometer:            dbVehicle.Odometer,
+				Speed:               dbVehicle.Speed,
+				CongestionLevel:     dbVehicle.CongestionLevel,
+				UpdatedAt:           dbVehicle.UpdatedAt,
+				CurrentStopSequence: dbVehicle.CurrentStopSequence,
+				OccupancyStatus:     dbVehicle.OccupancyStatus,
+				OccupancyPercentage: dbVehicle.OccupancyPercentage,
+				StopID:              dbVehicle.StopID,
+				StopName:            dbVehicle.StopName,
+				TripID:              dbVehicle.TripID,
+				TripDirectionID:     dbVehicle.TripDirectionID,
+				RouteID:             dbVehicle.RouteID,
+				RouteColor:          dbVehicle.RouteColor,
```

Comment:
This is one inconvenient thing about using sqlc, which is that even if queries return the same columns, different Go types are generated and this mapping this needs to happen. The exception to this rule is if you query just a whole table (say `SELECT * FROM vehicle`) in which case you always get a `Vehicle` return type. For this reason I've generally preferred queries that just query a whole table. Nothing actionable in this comment, just an observation!

Reply:
Yeah, I remember spending some time tying to work around this haha. I am surprised sqlc doesn't have a way to explicitly merge types in these cases.